### PR TITLE
Fixed a three problems that could happen when you have bundle products

### PR DESCRIPTION
**Pull request by ResuBaka**
### Related issues

closes #

### Short description and why it's useful

1. Is that the optionChanged event does not work because camel case does not work 100% of the time.
2. Bundles with only one option for the bundle option were not working as the function find does not exist on an object.
3. This can only happen in an bundle or config product and that is that there is no server_item_id or item_id in the products that are going to be compared and then they where equal which is wrong.

All three cases above are fix with this PR.

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [x] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and currently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/themes/default/components/core/ProductBundleOptions.vue [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <template>
   <form class="custom-options">
     <div v-for="option in product.bundle_options" :key="('bundleOption_' + option.option_id)">
-      <product-bundle-option :option="option" @optionChanged="optionChanged" :error-messages="errorMessages" />
+      <product-bundle-option :option="option" @option-changed="optionChanged" :error-messages="errorMessages" />
     </div>
```
Comment (pkarw) [inline src/themes/default/components/core/ProductBundleOptions.vue]: Shouldn’t it be “bundleOptionChanged” as for the function name?

File: core/modules/cart/helpers/productsEquals.ts [hunk 2]
```diff
@@ -15,8 +15,12 @@ const getProductType = (product: CartItem): string =>
 const getServerItemId = (product: CartItem): string | number =>
   product.server_item_id || product.item_id
 
-const isServerIdsEquals = (product1: CartItem, product2: CartItem): boolean =>
-  getServerItemId(product1) === getServerItemId(product2)
+const isServerIdsEquals = (product1: CartItem, product2: CartItem): boolean => {
+  if (getServerItemId(product1) === undefined || getServerItemId(product2) === undefined) {
+    return false
```
Comment (pkarw) [inline core/modules/cart/helpers/productsEquals.ts]: What if both products have empty server_item_id (say both products added in the offline mode and not yet synced); it’s probably from the caller context that it should be false
In this case correct?

File: core/modules/cart/helpers/productsEquals.ts [hunk 3]
```diff
@@ -15,8 +15,12 @@ const getProductType = (product: CartItem): string =>
 const getServerItemId = (product: CartItem): string | number =>
   product.server_item_id || product.item_id
 
-const isServerIdsEquals = (product1: CartItem, product2: CartItem): boolean =>
-  getServerItemId(product1) === getServerItemId(product2)
+const isServerIdsEquals = (product1: CartItem, product2: CartItem): boolean => {
+  if (getServerItemId(product1) === undefined || getServerItemId(product2) === undefined) {
```
Comment (andrzejewsky) [inline core/modules/cart/helpers/productsEquals.ts]:
```suggestion
  const product1ItemId = getServerItemId(product1)
  const product2ItemId = getServerItemId(product2)
  const areItemIdsDefined = product1ItemId !== undefined && product2ItemId !== undefined
 
 return areItemIdsDefined && product1ItemId === product2ItemId
```

File: core/modules/catalog/components/ProductBundleOption.ts [hunk 4]
```diff
@@ -29,7 +29,10 @@ export const ProductBundleOption = {
       return `bundleOptionQty_${this.option.option_id}`
     },
     value () {
-      return this.option.product_links.find(product => product.id === this.productOptionId)
+      if (Array.isArray(this.option.product_links)) {
```
Comment (andrzejewsky) [inline core/modules/catalog/components/ProductBundleOption.ts]:
```suggestion
      const { product_links } = this.option
      if (Array.isArray(product_links)) {
        return product_links.find(product => product.id === this.productOptionId)
      }
      
      return product_links
```

File: core/modules/catalog/components/ProductBundleOption.ts [hunk 5]
```diff
@@ -57,13 +60,18 @@ export const ProductBundleOption = {
   methods: {
     setDefaultValues () {
       if (this.option.product_links) {
-        const defaultOption = this.option.product_links.find(pl => { return pl.is_default })
+        let defaultOption
+        if (Array.isArray(this.option.product_links)) {
+          defaultOption = this.option.product_links.find(pl => { return pl.is_default })
+        } else {
+          defaultOption = this.option.product_links
+        }
```
Comment (andrzejewsky) [inline core/modules/catalog/components/ProductBundleOption.ts]:
```suggestion
      const { product_links } = this.option
      const defaultOption = Array.isArray(product_links) ? product_links.find(pl => pl.is_default) : product_links
```

File: src/themes/default/components/core/ProductBundleOptions.vue (hunk 1 above)
Comment (ResuBaka) [reply src/themes/default/components/core/ProductBundleOptions.vue]: I think that this is fine as it is. As it only Happens in files that have bundle in the name so it should be clear what is happening here.

When you think it is not so clear we can change it.

File: core/modules/cart/helpers/productsEquals.ts (hunk 2 above)
Comment (ResuBaka) [reply core/modules/cart/helpers/productsEquals.ts]: I should always be false when one of the products has no server_item_id or item_id other wise when both don't have it they are the same product but aren't in most cases.

File: core/modules/catalog/components/ProductBundleOption.ts (hunk 5 above)
Comment (ResuBaka) [reply core/modules/catalog/components/ProductBundleOption.ts]: Normally I would write it like this

```typescript
const defaultOption = Array.isArray(product_links) ? product_links.find(pl => {
          return pl.is_default
        }) : product_links
```

as I try to always be under the 120 length in a line. 

What do you think about that?

File: core/modules/cart/helpers/productsEquals.ts (hunk 3 above)
Comment (ResuBaka) [reply core/modules/cart/helpers/productsEquals.ts]: Done

File: core/modules/catalog/components/ProductBundleOption.ts (hunk 4 above)
Comment (ResuBaka) [reply core/modules/catalog/components/ProductBundleOption.ts]: Done

File: core/modules/cart/helpers/productsEquals.ts (hunk 2 above)
Comment (ResuBaka) [reply core/modules/cart/helpers/productsEquals.ts]: But for that case we have the check of the checksum.

File: core/modules/catalog/components/ProductBundleOption.ts (hunk 5 above)
Comment (andrzejewsky) [reply core/modules/catalog/components/ProductBundleOption.ts]: yeah it's ok, but you can break line in the `inline if` just after question mark or colon eg:
```
condition ? // wrap line
 true : // wrap line
 false
```

Comment (andrzejewsky) [reply core/modules/catalog/components/ProductBundleOption.ts]: also you don't need brackets here:

```
product_links.find(pl => { return pl.is_default }) -> product_links.find(pl =>pl.is_default)
```

Comment (ResuBaka) [reply core/modules/catalog/components/ProductBundleOption.ts]: I have changed it I can do it 100% like you want it because of eslint :smile:

Comment (andrzejewsky) [reply core/modules/catalog/components/ProductBundleOption.ts]: @ResuBaka ESlint always knows better how to make you happier 😆

**Comment by andrzejewsky** [comment]
@ResuBaka could you please add a short info how to test / reproduce bug you mentioned?  - this is for understanding for QA team.

**Comment by ResuBaka** [comment]
As it is already merged I still write in here how you could have run into some of the problems.


The Event name change is that in our theme it has not worked I didn't under stand why. So I changed it to more like how actual dom event are named as there are no uppercase in them allowed as far as I am aware of it. 

Then the change in the for checking if an bundle option has multible options is when each or one bundle option has only one option in it.

The change for if an product has an server_item_id or item_id is there when it happens that both are not set for what ever reason in both product both are undefined that it would be true which is wrong. This is hard to force I thing as I don't even know what that is happen for us but as it is an edge case it is still good to have it fixed.

**Pull request merged**